### PR TITLE
Add named queuing system

### PR DESCRIPTION
**Pull request by octacian**
* Add `core.enqueue` and `core.dequeue`
* Rename `builtin/common/after.lua` to `queue.lua`

This named queuing system works like `core.after` but allows jobs to be referenced by name. What makes this useful, is how jobs are referenced by name, meaning that you can at any point remove a particular job or modify a particular job, by simply reassigning the job with `core.enqueue`.

The use case for this is mods that introduce a HUD, which you would like to hide after a certain number of seconds, for example, a HUD to display the name of the item you are holding. However, whenever you change the item you are holding, you want to move the time after which the HUD disappears back from when the HUD was changed. With `minetest.after`, you would end up with jobs stacking up, each one hiding the HUD even though it only needed to be hidden once. This left a globalstep or complex check inside `minetest.after` as the only possibilities, however, `core.enqueue` allows you to override old jobs, meaning that jobs never stack up.

Example Code:
```lua
local hello = "world"
core.queue("test", 5, function()
	minetest.chat_send_all("Hello "..hello.."!")
end, hello)

core.queue("test", 7, function()
	minetest.chat_send_all("Hello again!")
end)

-- Prevents job "test" from being run
core.dequeue("test")
```

The above code adds a job `test` set to expire in 5 seconds, which is instead overridden to expire in 7 seconds and print different text.

If I need to make this into separate commits, one for rename and one for add of queuing system, I can. However, nothing related to the way after itselfs functions was modified, only the check for `if #jobs < 1 then return end` was changed to `if #jobs > 0 then ... end`.

### CODE REVIEW

File: builtin/common/queue.lua [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		func = func,
+		expire = time + after,
+		arg = {...},
+		mod_origin = core.get_last_run_mod(),
+	}
+end
+
+function core.enqueue(name, after, func, ...)
+	assert(name and tonumber(after) and type(func) == "function",
+			"Invalid core.queue invocation")
```
Comment (SmallJoker) [inline builtin/common/queue.lua]: * `core.enqueue`

File: doc/lua_api.txt [hunk 2]
```diff
@@ -2800,6 +2800,12 @@ These functions return the leftover itemstack.
 * `minetest.after(time, func, ...)`
     * Call the function `func` after `time` seconds, may be fractional
     * Optional: Variable number of arguments that are passed to `func`
+* `minetest.queue(name, time, func, ...)`
+		* Call the function `func` after `time` seconds, may be fractional
+		* Jobs are referred to be `name`, and can be modified by calling this function again with the same name
```
Comment (SmallJoker) [inline doc/lua_api.txt]: \* `core.enqueue`

Also wrong indents for the function description.

File: doc/lua_api.txt [hunk 3]
```diff
@@ -2800,6 +2800,12 @@ These functions return the leftover itemstack.
...
+		* Optional: Variable number of arguments that are passed to `func`
+* `minetest.dequeue(name)`
+		* Remove the queue job `name` previously added with `minetest.queue` to prevent it from running
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Again `minetest.enqueue`

**Comment by octacian** [comment]
@SmallJoker I had a couple minutes, so, fixed the issues you mentioned.

**Comment by octacian** [comment]
BTW, the TravisCI build failure does not seem to be related to this PR, but apparently some issue with "undefined symbols" in the area of network packets.

**Comment by nerzhul** [comment]
@octacian please rebase

**Comment by octacian** [comment]
@nerzhul Done. I assume the rebase should solve the TravisCI issue? Cause there were no merge conflicts, so...

**Comment by rubenwardy** [comment]
I'd quite like this to be implemented as part of an event system.

```lua
minetest.register_event("test",
    persist = true,
    func = function(one, two)
        minetest.chat_send_all(one .. ", " .. two)
    end,
})


-- elsewhere
minetest.start_timer("test", "one", "two")
```

Or similar. One uniform API for both this PRs usecase, and persisting timers usecase. Otherwise we end up with 3 different APIs :(

**Comment by octacian** [comment]
I like that idea. If no one else is able, I'd be happy to try to implement it when I get back to Minetest next month. My assumption would be that we would leave the basic `minetest.after` code as is and implement this separately in the file where the `minetest.register_*` code is found? Now, I assume that this couldn't be done entirely in Lua, as at least the persisting of events would have to be handled by C++, but I'd still be happy to give it a go later.

**Comment by paramat** [comment]
Needs work but neglected, can be reopened when there is progress.

**Comment by octacian** [comment]
Should I finish this PR or drop it in favour of @rubenwardy's suggestion? Such an events system would be much more powerful, however, this might still be worthwhile due to it having a simpler API.

For example, adding something to the queue with this PR can be done with proper formatting in as little as 3 lines of code. The same goes for overriding it. With rubenwardy's suggestion, it would take at least 4 lines of code to achieve the same thing, and a separate API call would be necessary to stop/override timers.

I guess that additional complexity is honestly negligible. So yeah, I'm good either way.

**Pull request closed without merging**